### PR TITLE
BUGFIX: Limit click-outside handling for dialogs to their semi-transparent overlay

### LINKED ISSUE #2925: Remove click-outside handling for Dialog
### Motivation

I've come across multiple scenarios (namely: https://github.com/sitegeist/Sitegeist.Archaeopteryx and https://github.com/Flowpack/media-ui) in which UI extensions need to stack dialogs on top of each other.

This interferes with our current click-outside handling for Dialogs, which makes the assumption that such stacking never happens.

The effect of that can be seen here:

https://user-images.githubusercontent.com/2522299/126071348-86f89542-a69e-40ca-b5bb-d4edd5a3b8c5.mp4

You see that the new Media UI is embedded in the Node Creation Dialog (Dialog 1). Then we click on "Upload", which opens the Upload Modal of the new Media UI (Dialog 2). If we click anywhere inside Dialog 2, both dialogs close. When we click ouside Dialog 2, both dialogs close as well.

https://user-images.githubusercontent.com/2522299/126071499-f95ad2ee-90cc-4a2e-b034-901ee4aa3f5f.mp4

The reason is, that after opening Dialog 2, the entire screen is now an area that is considered to be outside Dialog 1. So if we click anywhere, Dialog 1 receives its request to close. Since Dialog 1 is removed from VDOM, and Dialog 2 is a child of Dialog 1, Dialog 2 has to leave as well.

### Possible Solution

Imho, we don't need such aggressive click-outside handling for our Dialogs. It would be sufficient to just react to clicks on the semi-transparent overlay, since it covers (almost) the entire screen.

I say almost, because the top bar of the ui is not covered:

Screenshot_2021-07-18_16-47-08

This can potentially be worked around, because Dialogs live in an isolated part of the DOM tree, but I don't know if this behvaior is even intentional in the first place :sweat_smile:. (At least it doesn't seem right to me :monocle_face:)

I'd suggest to implement a solution as a bugfix for 5.3+.

### Affected Versions

Neos: >=3.3
UI: >=1.0

### PR DESCRIPTION
**Pull request by grebaldi**
fixes: #2925 

**The problem**

As documented in #2925, our click-outside-handling for dialogs is a tiny bit too aggressive, which leads to problems with stacked dialogs.

**The solution**

I removed the `enhanceWithClickOutside` decorator from the `<Dialog/>`-component and replaced its behavior with a simple `onClick`-handler that is limited to the semi-transparent overlay that surrounds every dialog.

The other issue described in #2925 (overlays do not cover entire user interface) is being addressed over here: https://github.com/neos/neos-ui/pull/3491

**EDIT:** Some more changes were necessary to apply the same behavior to when the escape key gets pressed. A detailed description of the reasoning behind those changes can be found in this comment: https://github.com/neos/neos-ui/pull/3492#issuecomment-1587405388

### CODE REVIEW

**Review by crydotsnake — CHANGES_REQUESTED** [review]
It works great with mouse clicks! But if i press the escape (esc) key on my keyboard, it closes all overalys. Can we intercept the escape key as well?

**Comment by grebaldi** [comment]
Very good point! I'll take a look at that :)

**Comment by grebaldi** [comment]
Well, interesting: I looked at the Escape-Key handling for the dialogs and found that since 2018 `react-close-on-escape` was supposed to be replaced by a vanilla solution (see: https://github.com/neos/neos-ui/pull/2274).

Yet, somehow `react-close-on-escape` managed to sneak back in. I'll try to restore the intent of #2274 within this PR, but as of yet I have no idea what happened there.

<s>Maybe it has something to do with the large esbuild change (but that's a wild, wild guess).</s>
**EDIT:** I wasn't thinking straight here... esbuild cannot be the cause, since we're on 7.3 :facepalm: 

/cc @mhsdesign @Sebobo @markusguenther

**Comment by mhsdesign** [comment]
you deserve a medal for https://github.com/neos/neos-ui/pull/3492/commits/b13d470cbceda5636aad5128af143bc49924e703 ;)

we definitely make sure that this doesnt happen with esbuild (despite the different import syntax, but didnt think the webpack stack had any obvious flaws like this one ^^)

-----------


Some modules referred to the components via

```js
import { /* ... */  } from '@neos-project/react-ui-components';
```

This caused webpack to pull the components from
'@neos-project/react-ui-components/lib-esm' rather than
'@neos-project/react-ui-components/src', leading to the
component modules being bundled twice.

The solution was to change those reference to an explicit:

```js
import { /* ... */  } from '@neos-project/react-ui-components/src';
```

Now the `react-ui-components` are only bundled once.

**Comment by grebaldi** [comment]
This was more than I thought :)

My plan was to have some global entity that can ensure a global order of stacked dialogs and close them in LIFO order when the escape-key is hit.

Plan didn't work at first try, because it turned out that webpack was bundling the react-components twice. Plugins would receive a different version of the components than the main UI - thus also a different version of the global stack-management entity.

So, I looked into that... Problem was that some modules in the codebase referred to the components via

```js
import { /* ... */  } from '@neos-project/react-ui-components';
```

This caused webpack to pull the components from '@neos-project/react-ui-components/lib-esm' rather than '@neos-project/react-ui-components/src', leading to the component modules being bundled twice.

The solution was to change those references to an explicit:

```js
import { /* ... */  } from '@neos-project/react-ui-components/src';
```

Now the `react-ui-components` are only bundled once.

With that, I introduced the `DialogManager` class, which takes care of managing the closing behavior of dialogs when the escape key is pressed.

I then proceeded to also remove `react-close-on-escape` as well as `react-portal` from the dialog component. The former because it wasn't supposed to be there in the first place and the latter because `react-dom` has its own Portal-API, so `react-portal` is no longer needed.

After this, `react-portal` is only used in `@neos-project/neos-ui` somewhere in the Inspector implementation. Removing it there should be done in a separate PR though.

Long story short: 
> But if i press the escape (esc) key on my keyboard, it closes all overalys. Can we intercept the escape key as well?

Yep, that works now :D

**Review by crydotsnake — APPROVED** [review]
Thanks a lot for the improvements! Everything works as expected 💙

File: packages/neos-ui-ckeditor5-bindings/src/EditorToolbar/LinkButton.js [hunk 1]
```diff
@@ -4,7 +4,7 @@ import {connect} from 'react-redux';
 import {$get, $transform} from 'plow-js';
 import LinkInput from '@neos-project/neos-ui-editors/src/Library/LinkInput';
 
-import {IconButton} from '@neos-project/react-ui-components';
+import {IconButton} from '@neos-project/react-ui-components/src';
```
Comment (mhsdesign) [inline packages/neos-ui-ckeditor5-bindings/src/EditorToolbar/LinkButton.js]: ill see if i can fix this differently, maybe via webpack alias?

that would have the benefit, that people dont need to know how to avoid bundling the components twice.
Also we wont disturb our efforts in: https://github.com/neos/neos-ui/pull/3465

**Review by mhsdesign — CHANGES_REQUESTED** [review]

File: packages/neos-ui-ckeditor5-bindings/src/EditorToolbar/LinkButton.js (hunk 1 above)
Comment (mhsdesign) [reply packages/neos-ui-ckeditor5-bindings/src/EditorToolbar/LinkButton.js]: in theory this should work:

in our [root webpack config](https://github.com/neos/neos-ui/blob/f042f183858b56743465b5806ee978a89b8c001f/webpack.config.js#L28) add this line:

```js
  ...
  resolve: {
    alias: {
      '@neos-project/react-ui-components$': path.resolve(__dirname, './packages/react-ui-components/src'),
      // idk if the next line is needed
      '@neos-project/react-ui-components/$': path.resolve(__dirname, './packages/react-ui-components/src')
    },
  },
```

it would be cool if someone could try this out and revert https://github.com/neos/neos-ui/commit/b13d470cbceda5636aad5128af143bc49924e703

maybe you? @crydotsnake :D xD

(i dont have a webpack build setup atm ^^ and i am to proud to change that. esbuild is my baby)

Comment (grebaldi) [reply packages/neos-ui-ckeditor5-bindings/src/EditorToolbar/LinkButton.js]: Oh, yep, that looks about right. Didn't think of that :) I'll test this.

Comment (mhsdesign) [reply packages/neos-ui-ckeditor5-bindings/src/EditorToolbar/LinkButton.js]: thanks for giving it a shot ;)
dont worry that this is ugly af. I have seen the sun. Its only a matter of time until esbuild will totally outlive the webpack buildstack. So im fine whatever we do with the old one ^^ It doesnt matter.

Comment (grebaldi) [reply packages/neos-ui-ckeditor5-bindings/src/EditorToolbar/LinkButton.js]: Works like a charm :) Thanks for the pointer @mhsdesign!

> // idk if the next line is needed

It's not :)

**Comment by mhsdesign** [comment]
Thanks for the adjustments @grebaldi ;) I will try this out locally soon.

of topic a bit, but what do you think about https://github.com/neos/neos-ui/issues/3531, linking it here because its related.
We can continue discussing that over there ^^

File: packages/neos-ui-ckeditor5-bindings/src/EditorToolbar/LinkButton.js (hunk 1 above)
Comment (mhsdesign) [reply packages/neos-ui-ckeditor5-bindings/src/EditorToolbar/LinkButton.js]: Thank you for the implementation ;)

File: packages/react-ui-components/src/Dialog/DialogManager.ts [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+    removeEventListener: Document['removeEventListener'];
+}
+
+export interface Dialog {
+    close: () => void;
+}
+
+/**
+ * @TODO
+ * Replace this class with a proper dialog management solution
```
Comment (mhsdesign) [inline packages/react-ui-components/src/Dialog/DialogManager.ts]: i think the implementation is _proper_ enough. Just want to guard against todo's that we aint gonna solve (we have already a lot of them and some with higher priority)

Comment (grebaldi) [reply packages/react-ui-components/src/Dialog/DialogManager.ts]: Can't argue that :D

Makes sense, I removed the comment.

File: packages/react-ui-components/src/Dialog/DialogManager.spec.ts [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+            expect(eventRoot.removeEventListener).not.toBeCalled();
+
+            dialogManager.closeLatest();
+
+            expect(eventRoot.removeEventListener).toBeCalledWith(
+                'keydown',
+                dialogManager.handleKeydown
+            );
+
+            dialogManager.closeLatest();
```
Comment (mhsdesign) [inline packages/react-ui-components/src/Dialog/DialogManager.spec.ts:171]:
```suggestion
```

noop

File: packages/react-ui-components/src/Dialog/DialogManager.ts [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+            this.closeLatest();
+        }
+    }
+
+    public closeLatest(): void {
+        const dialog = this.dialogs.pop();
+        if (dialog) {
+            dialog.close();
+            this.removeHandleKeydownEventListenerIfNecessary();
+        }
```
Comment (mhsdesign) [inline packages/react-ui-components/src/Dialog/DialogManager.ts:41]:
```suggestion
        const lastDialog = this.dialogs[this.dialogs.length - 1];
        if (lastDialog) {
            this.forget(lastDialog);
        }
```

as `closeLatest` is only used for testing, id rather have it cover the `forget` method aswell ^^

**Comment by bwaidelich** [comment]
Sorry for being too lazy to test, but will this also affect the behavior of non-stacked dialogs?
So often I fill out a large creation dialog and close it by accident because I click next to it (or hit backspace to fix an input)

**Comment by grebaldi** [comment]
@bwaidelich 
I'm afraid not. I've seen that https://github.com/neos/neos-ui/issues/3531 has been reported (which I believe includes your description - except for the "backspace"-part). This PR only covers #2925, which came up during Media UI development.

I totally agree that #3531 needs to be addressed. Not sure whether this PR is the place to do so, though.

**Comment by bwaidelich** [comment]
> I'm afraid not. I've seen that https://github.com/neos/neos-ui/issues/3531 has been reported

Ah right, I didn't see that.
I'll try to reproduce that backspace-issue and comment it there, thanks!

**Review by mhsdesign — APPROVED** [review]
Thanks ;) Works as expected!
